adapters/mediago/mediago.go

```diff
@@ -0,0 +1,214 @@
+	imp := request.Imp[0]
+	err = json.Unmarshal(imp.Ext, &extBidder)
+	if err != nil {
+		return nil, err
+	}
+
+	var extImpMediaGo openrtb_ext.ExtImpMediaGo
+	err = json.Unmarshal(extBidder.Bidder, &extImpMediaGo)
+	if err != nil {
+		return nil, err
+	}
+	if extImpMediaGo.Token != "" {
+		extMediaGo.Token = extImpMediaGo.Token
+		extMediaGo.Region = extImpMediaGo.Region
+
+		return &extMediaGo, nil
+	}
+	return nil, errors.New("mediago token not found")
+
+}
+
+func getRegionInfo(region string) string {
+	switch region {
+	case "APAC":
+		return "jp"
+	case "EU":
+		return "eu"
+	case "US":
+		return "us"
+	default:
+		return "us"
+	}
+}
+
+func (a *adapter) getEndPoint(ext *openrtb_ext.ExtMediaGo) (string, error) {
+	endPointParams := macros.EndpointTemplateParams{
+		AccountID: url.PathEscape(ext.Token),
+		Host:      url.PathEscape(getRegionInfo(ext.Region)),
+	}
+	return macros.ResolveMacros(a.EndpointTemplate, endPointParams)
+}
+
+func preProcess(request *openrtb2.BidRequest) {
+	for i := range request.Imp {
+		if request.Imp[i].Banner != nil {
+			banner := *request.Imp[i].Banner
+			if (banner.W == nil || banner.H == nil || *banner.W == 0 || *banner.H == 0) && len(banner.Format) > 0 {
+				firstFormat := banner.Format[0]
+				banner.W = &firstFormat.W
+				banner.H = &firstFormat.H
+				request.Imp[i].Banner = &banner
+			}
+		}
+	}
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(response) {
+		return nil, nil
+	}
+	if err := adapters.CheckResponseStatusCodeForErrors(response); err != nil {
+		return nil, []error{err}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+	var errs []error
+
+	for _, seatBid := range bidResp.SeatBid {
+		for idx := range seatBid.Bid {
+			mediaType, err := getBidType(seatBid.Bid[idx], internalRequest.Imp)
+			if err != nil {
+				errs = append(errs, err)
+			} else {
+				bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+					Bid:     &seatBid.Bid[idx],
+					BidType: mediaType,
+				})
+			}
+		}
+	}
+
+	return bidResponse, errs
+}
+
+func getBidType(bid openrtb2.Bid, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+	switch bid.MType {
+	case openrtb2.MarkupBanner:
+		return openrtb_ext.BidTypeBanner, nil
+	case openrtb2.MarkupNative:
+		return openrtb_ext.BidTypeNative, nil
+	// case openrtb2.MarkupAudio:
+	// 	return openrtb_ext.BidTypeAudio, nil
+	// case openrtb2.MarkupVideo:
+	// 	return openrtb_ext.BidTypeVideo, nil
+	default:
+		for _, imp := range imps {
+			if imp.ID == bid.ImpID {
+				if imp.Banner != nil {
+					return openrtb_ext.BidTypeBanner, nil
+				}
+				if imp.Native != nil {
+					return openrtb_ext.BidTypeNative, nil
+				}
+			}
+		}
```

Comment:
any particular reason to fallback to this logic when no case for MType matches?

Reply:
> any particular reason to fallback to this logic when no case for MType matches?

Hi @onkarvhanumante chiming in here for my colleague @SylviaF . She replied to this question last week but somehow the reply was not seen. Essentially she's saying the reason why we map the MType field according to the bid type in imp is that we follow the standard OpenRTB specs, and MType is not included in oRTB specs. We will make modifications based on Prebid's requirements, however just in case, we want to keep this fallback code to make sure we can bid properly.

Reply:
@onkarvhanumante the Prebid Server committee discussed this thread the other day. We're ok with this fallback logic. Using the MType is preferable and there's no harm in having the fallback code. As I'm sure you're aware, a lot of other adapters only have the imp-based approach. I imagine you were just asking out of curiosity, but just in case, I want you to know that the committee is ok with this.